gateway-ha/src/main/java/com/lyft/data/gateway/ha/router/RuleReloadingRoutingGroupSelector.java

```diff
@@ -0,0 +1,84 @@
+package com.lyft.data.gateway.ha.router;
+
+import java.io.FileReader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.HashMap;
+import java.util.Optional;
+
+import javax.servlet.http.HttpServletRequest;
+
+import lombok.extern.slf4j.Slf4j;
+import org.jeasy.rules.api.Facts;
+import org.jeasy.rules.api.Rules;
+import org.jeasy.rules.api.RulesEngine;
+import org.jeasy.rules.core.DefaultRulesEngine;
+import org.jeasy.rules.mvel.MVELRuleFactory;
+import org.jeasy.rules.support.reader.YamlRuleDefinitionReader;
+
+@Slf4j
+public class RuleReloadingRoutingGroupSelector
+    implements RoutingGroupSelector  {
+
+  RulesEngine rulesEngine = new DefaultRulesEngine();
+  MVELRuleFactory ruleFactory = new MVELRuleFactory(new YamlRuleDefinitionReader());
+  String rulesConfigPath;
+  Rules rules = new Rules();
+  boolean rulesLoaded;
+  long lastUpdatedTime;
+
+  RuleReloadingRoutingGroupSelector(String rulesConfigPath) {
+    this.rulesConfigPath = rulesConfigPath;
+    try {
+      rules = ruleFactory.createRules(
+              new FileReader(rulesConfigPath));
+      BasicFileAttributes attr = Files.readAttributes(Path.of(rulesConfigPath),
+              BasicFileAttributes.class);
+      lastUpdatedTime = attr.lastModifiedTime().toMillis();
+
+      rulesLoaded = true;
+    } catch (Exception e) {
+      log.error("Error opening rules configuration file, using "
+              + "routing group header as default.", e);
+    }
+  }
+
+  @Override
+  public String findRoutingGroup(HttpServletRequest request) {
+    try {
+      BasicFileAttributes attr = Files.readAttributes(Path.of(rulesConfigPath),
+              BasicFileAttributes.class);
+      log.debug(String.format("Current modified time %s, last modified time %s",
+              attr.lastModifiedTime().toMillis(), lastUpdatedTime));
+      if (attr.lastModifiedTime().toMillis() > lastUpdatedTime) {
+        log.info(String.format("Updating rules to file modified at %s", attr.lastModifiedTime()));
+        synchronized (this) {
+          // TODO: talk to the guys about whether or not we should synchronize
+          //  on this, rules, or sth else
+          rules = ruleFactory.createRules(
```

Comment:
There is a potential race condition with rules. Assuming this method gets called by multiple threads, rules gets set synchronously, but then is read below outside thread protection... which means the value might be changed in between the write and read, might have an exception in between all that, etc.

Reply:
Really depends on what the rulesEngine is doing. You basically want to compute the routing group per thread (I think?). Perhaps sync on the method. How often is the method expected to be called? How long does this method take to execute? Sometimes we overestimate the hit on performance vs overall safety/ease of maintenance.

Reply:
My main concern here is preventing `ruleFactory.createRules` from being called by two threads at the same time. `ruleFactory.createRules` uses a yaml parser that uses global state, and parsing will fail if two threads begin updating that state. 

Thread 1 can check `attr.lastModifiedTime().toMillis() > lastUpdatedTime` and have the condition evaluate to false, but before it gets to `rulesEngine.fire(rules, facts)`, Thread 2 may find that condition evaluate to true, and update the rules. This is ok. The only problem would be if Thread 2 does something to cause `rulesEngine.fire` to fail. That is possible, so I'll add a try/catch.

In terms of performance impact, this method is called up to once per query and once per non-query HTTP request. It could be expensive to synchronize this method when used with Insights.